app/packages/state/src/hooks/useOnShiftClickLabel.ts

```diff
@@ -114,4 +199,120 @@ export const useOnShiftClickLabel = () => {
+        });
+
+        if (!similarLabels) {
+          return;
+        }
+
+        const currentHoveredInstances = jotaiStore.get(hoveredInstances);
+
+        if (!currentHoveredInstances || currentHoveredInstances.length !== 2) {
+          return;
+        }
+
+        const fieldName = (
+          currentHoveredInstances[1] as HoveredInstancesLabelsTuple[1]
+        )[sourceLabelId].field;
+
+        const { label_id_map: labelIdMap } = similarLabels;
+
+        const labelsToAddOrRemove = Object.entries(labelIdMap).map(
+          ([labelId, frameNumber]) => ({
+            sampleId,
+            labelId,
+            frameNumber,
+            field: fieldName,
+            instanceId: sourceInstanceId,
+          })
+        );
+
+        // if current label is already selected, and shift + click pressed,
+        // assume user wants to deselect all instances with that instance config
+        const currentSelectedLabels = snapshot
+          .getLoadable(selectedLabelMap)
+          .getValue();
+
+        if (currentSelectedLabels[sourceLabelId]) {
+          set(selectedLabels, (prev) => {
+            return [
+              ...prev.filter((label) => label.instanceId !== sourceInstanceId),
+            ];
+          });
+        } else {
+          set(selectedLabels, (prev) => {
+            return [...prev, ...labelsToAddOrRemove];
+          });
+        }
+      },
+    []
+  );
+
+  return useRecoilCallback(
+    ({ snapshot }) =>
+      async (sampleId: string, labels: LabelData[], e: LabelToggledEvent) => {
+        const { sourceInstanceId, sourceSampleId } = e.detail;
+
+        if (!sourceInstanceId) {
+          return;
+        }
+
+        /**
+         * IN GROUPS:
+         *
+         * Let k = number of currently hovered similar instances
+         * Let n = number of instances in the modal with that instance config
+         *
+         * Three possible cases:
+         * 1. If k = 0, shift + click selects all instances with that instance config
+         * 2. If k = n, shift + click deselects all instances with that instance config
+         * 3. If 0 < k < n, shift + click selects all instances with that instance config
+         *
+         * IN VIDEO DATASETS:
+         *
+         * Two possible cases:
+         * 1. Shift + click selects all instances with that instance config.
+         * 2. Shift + click deselects all instances with that instance config.
+         */
+
+        const hoveredSampleValue = snapshot
+          .getLoadable(hoveredSample)
+          .getValue() as unknown as FrameSample;
+
+        if (!hoveredSampleValue) {
+          return;
+        }
+
+        const isVideoWithMultipleFrames =
+          hoveredSampleValue?._media_type === "video" &&
+          hoveredSampleValue?.frames?.length > 0;
+
+        if (isVideoWithMultipleFrames) {
+          return handleVideo(sourceSampleId, labels, e);
+        } else {
+          return handleGroup(sampleId, labels, e);
+        }
```

Comment:
How does this feature work with a video slice in a group dataset? Or that is out of scope for now?

Reply:
yeah i'll look into it in next iteration.